File: matlab/calculateSim.m
function sim = calculateSim(x)
[row, ~] = size(x);
sim = zeros(row);

for i = 1:row
    for j = i + 1:row
        sim(i, j) = personSim(x(i, :), x(j, :));
    end
    fprintf("user %d calculate sim finished!\n", i);
end

sim = sim + sim' + eye(size(sim));

end

File: backend/recommend-thrift-python-service/recommend/util/CalculateSim.py
def cosSim(x):
    mod_x = (x * x.transpose())[0, 0]
    if mod_x == 0:
        return 1
    return 1 / (mod_x) ** 0.5

def calculateSim(data):
    row, column = data.shape
    sim = data * data.transpose()
    i = 0
    while i < row:
        mo = cosSim(data[i, :])
        sim[i, :] *= mo
        sim[:, i] *= mo
        i += 1
        print("calculate sim finished: {:.6%}".format(i / row))
    return sim
